---
title: "Converting_fly_genome"
author: "Isabel Birds"
date: "5 November 2020"
output: html_document
---
For converting genome fasta and gff files into files representitative of the transcriptome for use in ribosome profiling analysis using Riboviz. Currently desinged for fly genomes but I (3mma-mack) am planning on using a similar code to create transcriptome files for S. pombe so creating notes and understanding the major steps is a good start

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(Biostrings)
library(seqinr)
library(GenomicRanges) 
library(GenomicFeatures)
library(rtracklayer)
library(bedtoolsr)
library(dplyr)
options(bedtools.path = "/Users/cm13ijb/opt/anaconda3/bin")
set.seed(123)
```

Raw files:

* Dmel GFF, from ftp://ftp.flybase.net/releases/FB2020_04/dmel_r6.35/gff/dmel-all-r6.35.gff.gz 
* Dmel genome fasta, from ftp://ftp.flybase.net/releases/FB2020_04/dmel_r6.35/fasta/dmel-all-chromosome-r6.35.fasta.gz 

Need bedtools installed.
devtools::install_github("PhanstielLab/bedtoolsr")
Have to give
 options(bedtools.path = "/Users/cm13ijb/opt/anaconda3/bin")


To Do:

* Reformat dmel files to match yeast files - transcript focused. Previous script here https://github.com/riboviz/riboviz/blob/master/rmarkdown/script_for_transcript_annotation.Rmd - isn't generic.
* Check how bicistronic transcripts are dealt with in GFF. 
* Filter to one ORF per transcript at end (check - per gene?)
* Double check works on gencode and non flybase GFF - other species. 


Vague plan:

* Load files 
* Pull out CDS, UTRs from GFF
* Add/extend UTRs in GFF
* Use new GFF to get CDS and buffer sequence from fasta.
* Concatenate CDS in fasta
* Change fasta sequence IDs, and start/stops
* Make new GFF based on this fasta
* Output files

Notes:

* Dmel UTR length - based on current annotation - 5' UTR median 194nt, 3' UTR median 306nt. Some v v long outliers, but 250nt seems reasonable to use for buffer for now.
* GFF format aim: Seq_ID  X SeqType start end . strand  . Name=Seq_ID

naming convention:
five_prime_UTR
three_prime_UTR

___

Script requires - genome fasta file and GFF file. Length of flanking region - using approx median UTR length.

```{r user_inputs}
##change files to my required files - downloaded from pombase onto my computer 
annotation_path <- "1_Raw_data/Dmel_genomes/dmel-all-r6.35.gff"
genome_path <- "1_Raw_data/Dmel_genomes/dmel-all-chromosome-r6.35.fasta"
#How large do I want all 3' and 5' UTRs on either side of CDS to be
flanking_length_nt <- 250
# where do I want my end files to be stored  
output_path <- "2_Processed_data/Dmel_genomes/Dmel"
```

What is a TxDb object. a TxDb is made up of at least 2 dataframes (required are info on transcripts and info on splicings) with the options of a gene database. These dataframes 'describe a set of transcripts and the genomic features related to them (exons, CDS and genes at the moment)' and together form an object called a TxDb. This is from the GenomicFeatures package. 

Using make TxDbFromGFF as better for quickly loading large GFFs.
If also need exons: exons <- exonsBy(txdb,"tx")

```{r load_files} 

#import a GFF as a TxDb object. 
txdb <- GenomicFeatures::makeTxDbFromGFF(annotation_path, format="gff3")

#Extract the coding regions, UTRs, and exons by transcript

cds <- GenomicFeatures::cdsBy(txdb, "tx",use.names=TRUE)
threeUTR <-  GenomicFeatures::threeUTRsByTranscript(txdb,use.names=TRUE)
fiveUTR <- GenomicFeatures::fiveUTRsByTranscript(txdb,use.names=TRUE)

```

Background info while trying to understand data:

head(txdb) produces: 'Error in x[seq_len(n)] : object of type 'S4' is not subsettable' so was turned into a list to see the contents

test_txdb <- as.list(txdb) produces a list with 4 elements; $transcripts, $splicings, $genes, $chrominfo. to find out more each was explored with head().

> head(test_txbd$transcripts)
  tx_id       tx_name                tx_type             tx_chrom tx_strand tx_start tx_end
1     1  SPBC460.03.1                   mRNA chr_II_telomeric_gap         +    11641  13344
2     2  SPBC460.05.1                   mRNA chr_II_telomeric_gap         +    16470  18062
3     3 SPBC460.01c.1                   mRNA chr_II_telomeric_gap         -     1479   3197
4     4 SPBC460.02c.1                   mRNA chr_II_telomeric_gap         -     8856   9803
5     5 SPBC460.04c.1                   mRNA chr_II_telomeric_gap         -    14073  15248
6     6 SPAC212.09c.1 pseudogenic_transcript                    I         +     7619   9274

head(test_txbd$splicings)
  tx_id exon_rank exon_id            exon_name           exon_chrom exon_strand exon_start exon_end cds_id             cds_name cds_start cds_end cds_phase
1     1         1       1  SPBC460.03.1:exon:1 chr_II_telomeric_gap           +      11641    13344      1  SPBC460.03.1:exon:1     11641   13344         0
2     2         1       2  SPBC460.05.1:exon:1 chr_II_telomeric_gap           +      16470    18062      2  SPBC460.05.1:exon:1     16470   18062         0
3     3         1       3 SPBC460.01c.1:exon:1 chr_II_telomeric_gap           -       1479     3197      3 SPBC460.01c.1:exon:1      1479    3197         0
4     4         1       5 SPBC460.02c.1:exon:1 chr_II_telomeric_gap           -       9651     9803      5 SPBC460.02c.1:exon:1      9651    9803         0
5     4         2       4 SPBC460.02c.1:exon:2 chr_II_telomeric_gap           -       8856     9365      4 SPBC460.02c.1:exon:2      8856    9365         0
6     5         1       6 SPBC460.04c.1:exon:1 chr_II_telomeric_gap           -      14073    15248      6 SPBC460.04c.1:exon:1     14073   15248         0

head(test_txbd$genes)
  tx_id gene_id
1     1    vba2
2    14    ftm1
3    15    ftm2
4    20    ayr2
5    21    nhe1
6    22   prl19

head(test_txbd$chrominfo)
                 chrom length is_circular
1 chr_II_telomeric_gap     NA          NA
2                    I     NA          NA
3                   II     NA          NA
4                  III     NA          NA
5   mating_type_region     NA          NA
6        mitochondrial     NA          NA

the result of producing cds is a Large CompressedGRangesList (7006 elements, 1.8 MB) 

head(cds) produces: GRangesList object of length 6:
$SPBC460.03.1
GRanges object with 1 range and 3 metadata columns:
                  seqnames      ranges strand |    cds_id            cds_name exon_rank
                     <Rle>   <IRanges>  <Rle> | <integer>         <character> <integer>
  [1] chr_II_telomeric_gap 11641-13344      + |         1 SPBC460.03.1:exon:1         1
  -------
  seqinfo: 6 sequences from an unspecified genome; no seqlengths

$SPBC460.05.1
GRanges object with 1 range and 3 metadata columns:
                  seqnames      ranges strand |    cds_id            cds_name exon_rank
                     <Rle>   <IRanges>  <Rle> | <integer>         <character> <integer>
  [1] chr_II_telomeric_gap 16470-18062      + |         2 SPBC460.05.1:exon:1         1
  -------
  seqinfo: 6 sequences from an unspecified genome; no seqlengths

$SPBC460.01c.1
GRanges object with 1 range and 3 metadata columns:
                  seqnames    ranges strand |    cds_id             cds_name exon_rank
                     <Rle> <IRanges>  <Rle> | <integer>          <character> <integer>
  [1] chr_II_telomeric_gap 1479-3197      - |         3 SPBC460.01c.1:exon:1         1
  -------
  seqinfo: 6 sequences from an unspecified genome; no seqlengths

...
<3 more elements>

both head(threeUTR) and head(fiveUTR) produce 
'GRangesList object of length 0:
<0 elements>'

However, this may not be a problem as the next step adds flanks of 250nt to the 3' and 5' UTRs of genes in the next step. 

I wanted to create a smaller test file to run for the next step, however so far this has not been done. Will update here if successful. 


Describe UTRs for CDS without them.
This takes a while if no UTRs annotated.

```{r make_buffers}

#5' UTR
#Find CDS w/o 5' UTR
#line means record cds where the name present in cds does not appear in fiveUTR
no_fiveUTR_cds <- cds[!names(cds) %in% names(fiveUTR)]
# no_fiveUTR_cds when run with the full dataset is a Large CompressedGRangesList (7006 elements, 1.8 MB), the same as cds. This would explain why fiveUTR and threeUTR have a length 0, there are no recorded UTRs. This may be a problem, but will have to see later on

#Add 5'UTR
#this function will increase value in the ranges column of no_fiveUTR by a specified amount 
add_5UTR <- function(no_fiveUTR_cds,flanking_length_nt){
  #pull out start of cds
  #using to ensure retains all metadata and maps back to transcript.
  tmp_5UTR <- no_fiveUTR_cds[1]
  #find range for 5UTR,update range of tmp5_UTR
  ranges(tmp_5UTR) <- ranges(flank(tmp_5UTR,start = TRUE, both = FALSE, width = flanking_length_nt, use.names = TRUE))
  return(tmp_5UTR)
}
# when individually done, ranges() and flank() produce the error 'could not find function'. this does not happen when the whole function of add_5UTR() is run. Here it is used to indicate different columns in the data. 
no_fiveUTR_UTR <- endoapply(no_fiveUTR_cds,add_5UTR,flanking_length_nt)
#endoapply applies a data into a funtion, here add_5UTR. When run with the first 100 rows of no_fiveUTR_cds, the ranges increased by the specifeid amount (at the top of the code)


#3' UTR
#Find CDS w/o 3' UTR
no_threeUTR_cds <- cds[!names(cds) %in% names(threeUTR)]

#Add 3'UTR
add_3UTR <- function(no_threeUTR_cds,flanking_length_nt){
  #pull out end of cds
  #using to ensure retains all metadata and maps back to transcript.
  tmp_3UTR <- no_threeUTR_cds[length(no_threeUTR_cds)]
  #find range for 3UTR, update range of tmp3UTR
  ranges(tmp_3UTR) <- ranges(flank(tmp_3UTR,start = FALSE, both = FALSE, width = flanking_length_nt, use.names = TRUE))
  return(tmp_3UTR)
}

no_threeUTR_UTR <- endoapply(no_threeUTR_cds,add_3UTR,flanking_length_nt)
# this is a repeat of add_5UTR, and adds a flank to the 3' end of cds without a 3' UTR.
```

Extend UTRs of less than flanking length (250nt).
This bit takes a while... should depend on transcriptome size.
# I had run up to here by creating smaller datasets with the first 1000 rows of no_fiveUTR_cds and no_threeUTR_cds however I realised this may cause issues when I later came to recreate the GFF file. I may test run this with the full dataset soon
```{r extend_buffers}

#Find 5' UTR shorter than flanking length
short_5UTR <- fiveUTR[sum(width(fiveUTR)) < flanking_length_nt]

#function to extend 5'UTR
#ignores intron/exon boundaries currently
resize_short_5UTR <- function(short_5UTR,flanking_length_nt){
    #calculate extension needed
    extra = flanking_length_nt - sum(width(short_5UTR))
    #create extra range at start, combine with first range, update.
    #adds to start of first range for +, adds to end of first range for - strand
    ranges(short_5UTR[1]) <- ranges(range(flank(short_5UTR[1],start=TRUE,both=FALSE,width=extra),short_5UTR[1]))
    return(short_5UTR)
}
# very similar to add_fiveUTR, except only increasing flanks shorter than 250 to 250 instead of adding them on
extended_short_5UTR <- endoapply(short_5UTR,resize_short_5UTR,flanking_length_nt)


#Find 3' UTR shorter than flanking length
short_3UTR <- threeUTR[sum(width(threeUTR)) < flanking_length_nt]

#function to extend 3'UTR
#ignores intron/exon boundaries currently
resize_short_3UTR <- function(short_3UTR,flanking_length_nt){
    #calculate extension needed
    extra = flanking_length_nt - sum(width(short_3UTR))
    #create extra range at end, combine with last range, update.
    #adds to end of last range for +, adds to start of last range for - strand
    ranges(short_3UTR[length(short_3UTR)]) <- ranges(range(flank(short_3UTR[length(short_3UTR)],
                                                                 start=FALSE,both=FALSE,width=extra),short_3UTR[length(short_3UTR)]))
    return(short_3UTR)
}

extended_short_3UTR <- endoapply(short_3UTR,resize_short_3UTR,flanking_length_nt)
#after this all cds will have flanking regions of 250 nt on either side, this will be consistent throughout the data

#save outputs from this chunk because woof
save("extended_short_5UTR",file="tmp_extended_short_5UTR")
save("extended_short_3UTR",file="tmp_extended_short_3UTR")
# from base R, this function saves data as a new file, which is specifed in the file parameter 
```

Combine UTR list, check 5', 3' and CDS are same length. 
Combine all into one grangeslist.

```{r combine_UTR}

#Remove extended UTR from original three UTR - prevent duplicates. only threeUTR's that are not present in extender_short_3UTRs are kept in threeUTR
threeUTR <- threeUTR[!names(threeUTR) %in% names(extended_short_3UTR)]
#Combine into one three UTR list. all in this list should now have a flank of 250
all_threeUTR <- c(threeUTR,extended_short_3UTR,no_threeUTR_UTR) 

#Remove extended UTR from original five UTR - prevent duplicates.
fiveUTR <- fiveUTR[!names(fiveUTR) %in% names(extended_short_5UTR)]
#Combine into one five UTR list. repeat of above, all in this list should now have a length of 250
all_fiveUTR <- c(fiveUTR,extended_short_5UTR,no_fiveUTR_UTR)

#CHECK - should have a 3UTR and 5UTR for all CDS, and all flanking length or longer
length(cds) == length(all_threeUTR)
length(cds) == length(all_fiveUTR)
all((sum(width(all_threeUTR)) >= flanking_length_nt))
all((sum(width(all_fiveUTR)) >= flanking_length_nt))
# all() returns True or false. We are expecting true at this stage

#Clear up env. prevent confusion with having lots of documents 
rm(extended_short_3UTR,threeUTR,extended_short_5UTR,fiveUTR,no_fiveUTR_cds,no_fiveUTR_UTR,
   no_threeUTR_cds,no_threeUTR_UTR,short_3UTR,short_5UTR)
```

```{r get_sequences}

#Get rid of characters that will break R - EOF within quoted string error
names(cds) <- gsub("'","_",names(cds))
names(all_fiveUTR) <- gsub("'","_",names(all_fiveUTR))
names(all_threeUTR) <- gsub("'","_",names(all_threeUTR))
#gsub is a version of grep from base R. in this case it looks for the character ' and replaces it with _.

#Save cds and UTR as tmp bed files
rtracklayer::export.bed(cds,"tmp_cds.bed")
rtracklayer::export.bed(all_fiveUTR,"tmp_5UTR.bed")
rtracklayer::export.bed(all_threeUTR,"tmp_3UTR.bed")


#Use bedtools to get sequences. genome_path is the fasta file
#Split - splices together cds
cds_seq <- bedtoolsr::bt.getfasta(fi=genome_path, bed = "tmp_cds.bed",name = TRUE, split = FALSE)
three_seq <- bedtoolsr::bt.getfasta(fi=genome_path,bed = "tmp_3UTR.bed",name =TRUE, split = FALSE)
five_seq <- bedtoolsr::bt.getfasta(fi=genome_path,bed = "tmp_5UTR.bed",name =TRUE, split = FALSE)
#bt.getfasta 'Extract DNA sequences from a fasta file based on feature coordinates'.

#Check have expected no of sequences - make sure they are as expected
length(cds_seq$V1)/2 == length(cds)
length(five_seq$V1)/2 == length(cds)
length(three_seq$V1)/2 == length(cds)

#Tidy IDs, and add quote marks back. 
cds_seq$V1 <-  gsub(">","",cds_seq$V1)
cds_seq$V1 <-  gsub("_","'",cds_seq$V1)

five_seq$V1 <-  gsub(">","",five_seq$V1)
five_seq$V1 <-  gsub("_","'",five_seq$V1)

three_seq$V1 <-  gsub(">","",three_seq$V1)
three_seq$V1 <-  gsub("_","'",three_seq$V1)

```

Make new GFF based on fastas.
Three lines in GFF
YAL068C	rtracklayer	UTR5	1	250	.	+	.	Name=YAL068C
YAL068C	rtracklayer	CDS	251	613	.	+	.	Name=YAL068C
YAL068C	rtracklayer	UTR3	614	863	.	+	.	Name=YAL068C

```{r three_prime_UTR}
#Fastas are not read in same order
#Build dataframes for final GFF.
#five UTR
#cbind is from base R and combines multiple rows into a dataframe
tmp_five_gff <- cbind.data.frame(five_seq$V1[seq(1, nrow(five_seq), 2)],
                              five_seq$V1[seq(2, nrow(five_seq), 2)])
colnames(tmp_five_gff) <- c("fasta_id","five_UTR")
tmp_five_gff$type <- "five_prime_UTR"
tmp_five_gff$start <- 1
tmp_five_gff$width <- width(as.character(tmp_five_gff$five_UTR))
tmp_five_gff$end <- width(as.character(tmp_five_gff$five_UTR))

#CDS
tmp_cds_gff <- cbind.data.frame(cds_seq$V1[seq(1, nrow(cds_seq), 2)],
                              cds_seq$V1[seq(2, nrow(cds_seq), 2)])
colnames(tmp_cds_gff) <- c("fasta_id","cds")
tmp_cds_gff$type <- "CDS"
tmp_cds_gff <- left_join(tmp_cds_gff,tmp_five_gff %>% select(fasta_id,end),
                         by = "fasta_id")
tmp_cds_gff$start <- tmp_cds_gff$end + 1
tmp_cds_gff$width <- width(as.character(tmp_cds_gff$cds))
tmp_cds_gff$end <- tmp_cds_gff$end + width(as.character(tmp_cds_gff$cds))  
#left_join is from dplyr and is used for mutating the dataframe to join columns or rows

#three UTR
tmp_three_gff <- cbind.data.frame(three_seq$V1[seq(1, nrow(three_seq), 2)],
                              three_seq$V1[seq(2, nrow(three_seq), 2)])
colnames(tmp_three_gff) <- c("fasta_id","three_UTR")
tmp_three_gff$type <- "three_prime_UTR"
tmp_three_gff <- left_join(tmp_three_gff,tmp_cds_gff %>% select(fasta_id,end),
                         by = "fasta_id")
tmp_three_gff$start <- tmp_three_gff$end + 1
tmp_three_gff$width <- width(as.character(tmp_three_gff$three_UTR))
tmp_three_gff$end <- tmp_three_gff$end + width(as.character(tmp_three_gff$three_UTR))


#make gffs
#GRanges created a GRange object
five_gff <- GenomicRanges::GRanges(seqnames = tmp_five_gff$fasta_id,
                         ranges = IRanges(start = tmp_five_gff$start,
                                          width = tmp_five_gff$width),
                         strand = "+",
                         type = tmp_five_gff$type,
                         Name = tmp_five_gff$fasta_id)

cds_gff <- GenomicRanges::GRanges(seqnames = tmp_cds_gff$fasta_id,
                         ranges = IRanges(start = tmp_cds_gff$start,
                                          width = tmp_cds_gff$width),
                         strand = "+",
                         type = tmp_cds_gff$type,
                         Name = tmp_cds_gff$fasta_id)

three_gff <- GenomicRanges::GRanges(seqnames = tmp_three_gff$fasta_id,
                         ranges = IRanges(start = tmp_three_gff$start,
                                          width = tmp_three_gff$width),
                         strand = "+",
                         type = tmp_three_gff$type,
                         Name = tmp_three_gff$fasta_id)

#Combine into final gff, and export. Congrats you have a new gff3 file
final_gff <- c(five_gff,cds_gff,three_gff)

rtracklayer::export.gff3(final_gff,con = paste(output_path,".gff3",sep=""))

```

One sequence per CDS eg 
>YAL068C
ACCTATGAAAGATTTATGATTCGTTCAGAAACAAGAGCATCTCCATAGAGATAATGAGATTGTGTGAAAGATGAGATATA

```{r make final fasta}
#select the fasta_id and five_UTR from tmp_five_gff and add more fasta ids and sequences, forming a final output in the format five_UTR, cds, three_UTR
final_fasta <- select(tmp_five_gff,c("fasta_id","five_UTR"))
final_fasta <- left_join(final_fasta,tmp_cds_gff %>% select(fasta_id,cds),
                         by = "fasta_id")
final_fasta <- left_join(final_fasta,tmp_three_gff %>% select(fasta_id,three_UTR),
                         by = "fasta_id")
final_fasta$full_seq <- paste(final_fasta$five_UTR,final_fasta$cds,final_fasta$three_UTR,sep = "") 

#using function not mapply to ensure makes a new file
#write.fasta creates a fasta file from the given sequences 
for (i in (1:length(final_fasta$fasta_id))){
  if (i == 1){
    seqinr::write.fasta(sequences = final_fasta$full_seq[i], 
                names = final_fasta$fasta_id[i],
                file.out = paste(output_path,".fasta",sep=""),
                as.string = TRUE,
                open = "w")
  } else{
    seqinr::write.fasta(sequences = final_fasta$full_seq[i], 
                names = final_fasta$fasta_id[i],
                file.out = paste(output_path,".fasta",sep=""),
                as.string = TRUE,
                open = "a")
  }
}


```

Then remove tmp files!
this cleans up your work environment
congrats you made the files!




